### DISCUSSION TITLE
FluentUI v9 setup in Next.js 13 app directory

### DISCUSSION THREAD

**Discussion by iammathew** [microsoft/fluentui#25384] [Q&A]
Hey FluentUI community,

Next.js 13 got released yesterday and introduced a new way of routing (inside a new `/app` directory) which does not have a `_app.(js|ts)` and `_document.(js|ts)`. Unfortunately the migration guide does not mention how to port over a custom Document they just say: "Include global styles inside the root layout file".

So did anyone already try and/or succeed using FluentUI v9 using the new Next.js version? Or maybe can provide some pointers on how to achieve it, unfortunately my experience with SSR CSS during compile time solutions is quite limited :( 

Related issue in Griffel.js styling library: https://github.com/microsoft/griffel/issues/266

**Comment by AkiraVoid** [comment]
I am using next.js 13.0.3 and to be honest I don't even know the feature you mentioned, I just simply put _app.ts and _document.ts in pages folder and everything went OK.

**Reply by AkiraVoid** [reply]
Since it's a beta feature, maybe you have to wait for this feature becoming stable.

**Reply by AkiraVoid** [reply]
I don't know but I think that you could work with `Root layout`. v9 is basically:

1. Getting the initial styles in Document level by using `createDOMRenderer` and `renderToStyleElements`, then it will return render (created by `createDOMRenderer`) as props of App and insert initial styles into document head tag;
2. Rendering App with wrapping components `RendererProvider`, `SSRProvider` and `FluentProvider`.

Also, it seems like that in new version of next.js, the `Root layout` plays a role that previous played by _document and _app, because it is required to contain `<html>` and `<body>` tags. So I guess you could perform the 2 steps in this scope, and let wrapping components from v9 wrap all children in `<body>`.

**Reply by AkiraVoid** [reply]
Your `app/layout.tsx` will looks like this:
```tsx
import {
  createDOMRenderer,
  FluentProvider,
  RendererProvider,
  renderToStyleElements,
  SSRProvider,
  webLightTheme,
} from '@fluentui/react-components';

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  const renderer = createDOMRenderer();
  const styles = renderToStyleElements(renderer);

  return (
    <html lang='en'>
      <head>{styles}</head>
      <body>
        <RendererProvider renderer={renderer}>
          <SSRProvider>
            <FluentProvider theme={webLightTheme}>{children}</FluentProvider>
          </SSRProvider>
        </RendererProvider>
      </body>
    </html>
  );
}
```

**Reply by lszomoru** [reply]
I think that there is more to it. If you place the code above in `layout.tsx` you will get the following compilation error as the components in the new `app` directory as server-side by default. 

```
error - ./node_modules/@fluentui/react-dialog/lib/contexts/dialogSurfaceContext.js

You're importing a component that needs createContext. It only works in a Client Component but none of its parents are marked with "use client", so they're Server Components by default.
```

**Reply by AkiraVoid** [reply]
@lszomoru As document says, layouts can be set to client component, so we may could add a second level layout which contains those provider components?

I'm confused about `Server Components` and `Client Components`. Is that to say that children in `Client Components` can only be rendered within client side?

**Reply by lszomoru** [reply]
Based on [this documentation](https://beta.nextjs.org/docs/rendering/fundamentals) it looks like one can mix client/server components.

**Reply by AkiraVoid** [reply]
@lszomoru So we could:

```tsx
'use client'

import {
  FluentProvider,
  GriffelRenderer,
  RendererProvider,
  renderToStyleElements,
  SSRProvider,
  webLightTheme,
} from '@fluentui/react-components';

export default function FluentRenderer({
  children, renderer
}: {
  children: React.ReactNode;
  renderer: GriffelRenderer;
}) {
  const styles = renderToStyleElements(renderer);

  return (
        <RendererProvider renderer={renderer}>
          <SSRProvider>
            <FluentProvider theme={webLightTheme}>{children}</FluentProvider>
          </SSRProvider>
        </RendererProvider>
  );
}
```
Then:
```tsx

import {
  createDOMRenderer,
  webLightTheme,
} from '@fluentui/react-components';

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  const renderer = createDOMRenderer();
  const styles = renderToStyleElements(renderer);

  return (
    <html lang='en'>
      <head>{styles}</head>
      <body>
       <FluentRenderer renderer={renderer}>{children}</FluentRenderer>
      </body>
    </html>
  );
}
```

**Reply by lszomoru** [reply]
Yes, I believe that would work. But I have the feeling that something is still missing since it looks like *all* components that are using any FluentUI components would have to be client-side components. That would contradict [the documentation](https://react.fluentui.dev/?path=/docs/concepts-developer-server-side-rendering-nextjs-setup--page) which states that FluentUI v9 fully support server-side rendering.

**Reply by AkiraVoid** [reply]
From Next.js documentation, a Client Component is a component that is rendered on server and hydrated on client (to make it interactable), on the other hands, a Server Component is a component that is rendered on server and will not be hydrated on client. From me that's to say a Client Component is not a component that can only be client-side rendered, and this is so-called "fully support server-side rendering" in previous version of React (since before React 18, all components? will be hydrated on client side). Anyway, I am not familiar with these concepts, so I may have made a mistake on understanding.

**Comment by arcode** [comment]
I have got the same issue. I'm wondering why nobody from support team replies 😔

**Reply by AkiraVoid** [reply]
Check: https://github.com/microsoft/fluentui/issues/27227#issuecomment-1479325990

**Reply by arcode** [reply]
Thanks mate,

**Reply by dashkevych** [reply]
I make [web application](https://www.recitop.com/create) using Next.js and FluentUI, and also had issues with setup due to not very good docs. However, I was able to get it working and posted some [notes on how to set it up](https://www.tarascodes.com/nextjs-fluentui-setup).

However, I have encountered another issue regarding the bundle size. The initial loading time of the app is longer during development. Although the loading time is reduced in production, it is still not optimal.

It seems like a [common problem](https://github.com/vercel/next.js/discussions/17977). I tried to [modularize imports](https://medium.com/@under_the_hook/next-js-modularize-imports-687d7a2cddcf) but no luck. 

[Official FluentUI docs](https://react.fluentui.dev/?path=/docs/concepts-developer-build-time-styles--page) has some information regarding to optimize runtime but it does not very helpful as well.

**Reply by AkiraVoid** [reply]
@dashkevych Are you using `output:  standalone` of Next.js? It could significantly reduce the bundle size. And also make sure you only installed the production dependencies. The final bundle size of your app should be around 500mb (size of docker image, should be less if you don't use docker).

**Reply by dashkevych** [reply]
@AkiraVoid, thank you for suggestions. I am wondering if you experience a slow page load in localhost when first start the app (npm run dev) using FluentUI and Next.js (App Router)?

**Reply by AkiraVoid** [reply]
Actually even in page router. But I guess I know why. Fluent UI tries to insert many css styles and variables on loading. In page router, many of styles will be loaded on server side, but in app router, since FluentProvider is currently a client component, those styles may only can be loaded on client side. I think this causes the slow initial render.

**Reply by gaetandezeiraud** [reply]
@AkiraVoid  `Fluent UI tries to insert many css styles and variables on loading.` I have 1-2s of my page without style (only HTML) at loading. I think it is because the style is not yet write in the header. An idea how to fix that? Maybe a "loading screen"? In client side only.

**Reply by AkiraVoid** [reply]
@Brouilles If the whole page you write is rendered in client side, you have to wait your browser (or users have to wait their browsers)  to execute those JavaScript codes which must be executed **before** the CSS styles were inserted, such as loading the whole React framework. So, how many seconds you or your users will be stuck at the white screen depends on the performance of your or their machines, and how many codes you wrote or used before the actual rendering happens. This is a known caveat of client side rendering, and server side rendering is what we generally used to solve this issue, so I guess that you may have to move to server side rendering if you don't want this white screen.

**Reply by gaetandezeiraud** [reply]
Thank you sadly it is not possible because I use nextjs for the UI of a webview, for an app.
